packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +157,263 @@ if (process.env.NODE_ENV === 'development') {
+   *   i.e. the max guess should always be `≥` the correct `visibleCount`.
+   *   - With each such doubling, the new min guess is the current max guess (since underflow = we guessed low).
+   * - Set `visibleCount` to the `maxGuess`.
+   * - Repeat the following by calling `guessVisibleCount()` (keep re-rendering but not painting):
+   *   - Each time the container overflows, new max guess is the average of the two guesses.
+   *   - Each time the container does not overflow, new min guess is the average of the two guesses.
+   * - Stop when the average of the two guesses is the min guess itself. i.e. no more averaging possible.
+   * - The min guess is then the correct `visibleCount`.
+   */
+  const guessVisibleCount = React.useCallback(() => {
+    // If already stabilized, already guessing, or in unit test, do not guess.
+    if (isStabilized || isGuessing.current || isUnitTest) {
+      return;
+    }
+
+    try {
+      isGuessing.current = true;
+
+      // We need to wait for the ref to be attached so that we can measure available and required sizes.
+      if (containerRef.current == null) {
+        return;
+      }
+
+      // If there are no items, stabilize immediately.
+      if (itemsCount === 0) {
+        dispatch({ type: 'stabilize' });
+        return;
+      }
+
+      const dimension = orientation === 'horizontal' ? 'Width' : 'Height';
+      const availableSize = containerRef.current[`offset${dimension}`];
+      const requiredSize = containerRef.current[`scroll${dimension}`];
+
+      const isOverflowing = availableSize < requiredSize;
+
+      if (
+        // We have already found the correct visibleCount
+        (visibleCount === itemsCount && !isOverflowing) ||
+        // if the new average of visibleCountGuessRange will never change the visibleCount anymore (infinite loop)
+        (maxGuess - minGuess === 1 && visibleCount === minGuess)
+      ) {
+        dispatch({ type: 'stabilize' });
+        return;
+      }
+
+      // Before the main logic, the max guess MUST be >= the correct visibleCount for the algorithm to work.
+      // If not:
+      // - double the max guess and visibleCount: since we need to overflow.
+      // - set min guess to current visibleCount: since underflow means correct visibleCount >= current visibleCount.
+      if (maxGuess === visibleCount && !isOverflowing) {
+        dispatch({ type: 'shiftGuessRangeForward' });
+        return;
+      }
+
+      let newMinGuess = minGuess;
+      let newMaxGuess = maxGuess;
+
+      if (isOverflowing) {
+        // overflowing = we guessed too high. So, new max guess = half the current guess
+        newMaxGuess = visibleCount;
+      } else {
+        // not overflowing = maybe we guessed too low? So, new min guess = half of current guess
+        newMinGuess = visibleCount;
+      }
+
+      // Next guess is always the middle of the new guess range
+      const newVisibleCount = Math.floor((minGuess + maxGuess) / 2);
```

Comment:
I'm not sure that this calculation is right. It's still using the old `minGuess` and `maxGuess` values, so it wouldn't even change (since the old `visibleCount` likely was already in the middle).

Reply:
The only remaining suggestion I have is to split the `"update"` action into two.

```jsx
if (isOverflowing) {
  // overflowing = we guessed too high. So, decrease max guess
  dispatch({ type: 'decreaseMaxGuess' });
  return;
} else {
  // not overflowing = maybe we guessed too low? So, increase min guess
  dispatch({ type: 'increaseMinGuess' });
  return;
}
```

All the other code could be moved into the reducer, since we already have the old state values there.